qupid/matching.py

```diff
@@ -0,0 +1,79 @@
+import collections
+import random
+
+from networkx.algorithms.bipartite import sets as bipartite_sets
+
+INFINITY = float("inf")
+
+
+def hopcroft_karp_matching(G, top_nodes=None):
+    """Returns a maximum cardinality matching of the bipartite graph G.
+
+    NOTE: This function is modified from the NetworkX implementation:
+        https://tinyurl.com/akcfy6r4
+
+    We have modified the DFS algorithm to move to a *random* neighbor from all
+    possible neighbors. This avoids the issue where the original NetworkX
+    implementation was deterministic. We have also updated the names of some
+    variables for readability in case-control matching context.
+
+    :param G: Undirected bipartite graph
+    :type G: networkx.Graph
+
+    :param top_nodes: Container of nodes
+    :type top_nodes: Iterable
+
+    :returns: Dictionary matching each case to a single control
+    """
+    def breadth_first_search():
```

Comment:
Since there's a bfs and a dfs, do you need to randomize the order that nodes are pushed to the queue in the bfs as well?

Reply:
I don't think it's strictly necessary but probably would not hurt.